repository/src/main/java/org/apache/atlas/repository/store/aliasstore/ESAliasStore.java

```diff
@@ -198,13 +203,30 @@ private void personaPolicyToESDslClauses(List<AtlasEntity> policies,
                     }
 
                     for (String asset : assets) {
-                        if (asset.contains("*") || asset.contains("?")) {
-                            //DG-898 Bug fix
+                        /*
+                        * We are introducing a hierarchical filter for qualifiedName.
+                        * This requires a migration of existing data to have a hierarchical qualifiedName.
+                        * So this will only work if migration is done, upon migration completion we will set the feature flag to true
+                        * This will be dictated by the feature flag ENABLE_PERSONA_HIERARCHY_FILTER
+                        */
+
+                        // If asset resource ends with /* then add it in hierarchical filter
+                        boolean isHierarchical = asset.endsWith("/*");
+                        if (isHierarchical) {
+                            asset = asset.substring(0, asset.length() - 2);
+                        }
+                        boolean isWildcard = asset.contains("*") || asset.contains("?");
+                        if (isWildcard) {
```

Comment:
For asset value `default/redshift/1705171197/dev/public/*`
isHierarchical will be `true`
isWildcard will be `true`

Given this, it will never go to else block & hence `metadataPolicyQualifiedNames.add(asset);` will not be executed instead it will create `wildcard` as `isWildcard` is true


Ideally in this case `asset` should be added into `metadataPolicyQualifiedNames` (substring by -2) & no need of wildcard then

Reply:
Also, honestly, I feel this logic is complex to understand & maintain, and different check for useHierarchicalQualifiedNameFilter

can we separate them by useHierarchicalQualifiedNameFilter?

like 
```

boolean isWildcard = asset.contains("*") || asset.contains("?");

if (useHierarchicalQualifiedNameFilter) {
	if (isWildcard) {
		boolean isHierarchical = asset.endsWith("/*");

	    if (isHierarchical) {
		    asset = asset.substring(0, asset.length() - 2);
		    metadataPolicyQualifiedNames.add(asset);

		} else {
			allowClauseList.add(mapOf("wildcard", mapOf(QUALIFIED_NAME, asset)));
	        allowClauseList.add(mapOf("wildcard", mapOf(QUALIFIED_NAME, asset + "/*")));
		}
    } else {
    	metadataPolicyQualifiedNames.add(asset);
    }

} else {
	// this is as it is previous logic
	if (isWildcard) {
        allowClauseList.add(mapOf("wildcard", mapOf(QUALIFIED_NAME, asset)));
    } else {
	    terms.add(asset);
	}
	allowClauseList.add(mapOf("wildcard", mapOf(QUALIFIED_NAME, asset + "/*")));
}
```

Reply:
> For asset value default/redshift/1705171197/dev/public/*
> isHierarchical will be true
> isWildcard will be true

No `asset = asset.substring(0, asset.length() - 2);` will make sure we we cut `/*` from the string and consider it then. So isWildcard will be false.

Reply:
This code is like this because if we wanted to optimize for the case of `/*` in the end. If there is any we trim last part and proceed.

And added extra 2 conditions if we have hierarchical flag enabled, when we enable it for everyone the conditions will be simpler and easier to write code then. Its bit messy for the feature flag part actually. Otherwise its pretty simple.

Its already been tested a couple of time, I think we can proceed with this one.

Reply:
Understood